### PR TITLE
Merge the same items and refine a list of alternatives in error messages

### PR DESCRIPTION
**Pull request by arybczak**

### CODE REVIEW

**Comment by 23Skidoo** [comment]
You also need to update the `bootstrap.sh` script to install `edit-distance`.

**Comment by 23Skidoo** [comment]
And LambdaCase is not available in GHC 7.4.2, which we still support.

**Comment by arybczak** [comment]
@23Skidoo I updated the commit. Is it ok now?

**Comment by 23Skidoo** [comment]
@arybczak Let's wait until Travis is green. Patch itself looks OK to me, though it'd be nice with a before/after example in the commit message or PR text.

Do you think we can now simplify the code in https://github.com/haskell/cabal/blob/118c709754dc348b64910d050d389c3621e2e1aa/cabal-install/Distribution/Client/BuildTarget.hs#L698 ? At least the TODO can go now, I think.

/cc @dcoutts

**Comment by arybczak** [comment]
Right, this function can be simplified now.

Below is the example outputs before and after the change.

Before:
```
unknown@nihilist Cabal $ cabal new-build ansi-termnal
Resolving dependencies...
cabal: Unknown build target 'ansi-termnal'.
There is no component 'ansi-termnal' or component 'ansi-termnal'.
Perhaps you meant the component Cabal or unit-tests or package-tests or
QuickCheck or test-quickcheck or test-quickcheck-gcoarbitrary or
test-quickcheck-generators or test-quickcheck-gshrink or ansi-terminal or
ansi-wl-pprint
The project has no package 'ansi-termnal'.
Perhaps you meant the package Cabal or QuickCheck or ansi-terminal or
ansi-wl-pprint or cabal-install or clock or exceptions or hackage-security or
haskell-lexer or optparse-applicative
```

After:
```
unknown@nihilist Cabal $ ./cabal new-build ansi-termnal
Resolving dependencies...
cabal: Unknown build target 'ansi-termnal'.
There is no component 'ansi-termnal'.
Perhaps you meant the component ansi-terminal
The project has no package 'ansi-termnal'.
Perhaps you meant the package ansi-terminal
```

**Comment by dcoutts** [comment]
LGTM!

**Comment by 23Skidoo** [comment]
> ``` There is no component 'ansi-termnal'.
Perhaps you meant the component ansi-terminal
The project has no package 'ansi-termnal'.
Perhaps you meant the package ansi-terminal ```

Looks like there are still ways to improve our error reporting! Let's merge your patch once Travis is green, since it's already an improvement.

File: cabal-install/Distribution/Client/BuildTarget.hs [hunk 1]
```diff
@@ -498,15 +501,33 @@ resolveBuildTarget ppinfo opinfo userTarget =
...
                            <- map (innerErr Nothing) errs ]
-        nosuch   = [ (inside, thing, got, alts)
-                   | (inside, MatchErrorNoSuch thing got alts)
-                           <- map (innerErr Nothing) errs ]
+        nosuch   = Map.foldrWithKey genResults [] $ Map.fromListWith Set.union $
+          [ ((inside, thing, got), Set.fromList alts)
+          | (inside, MatchErrorNoSuch thing got alts) <- map (innerErr Nothing) errs
+          ]
+
+        genResults (inside, thing, got) alts acc = (
```
Comment (ezyang) [inline cabal-install/Distribution/Client/BuildTarget.hs]: Would be good to say what this chunk of code is trying to do. Something like, add a list of similarly named alternatives, giving up to three suggestions.

**Comment by ezyang** [comment]
sortOn was added in base 4.8.0.0, which is too early for us. I don't think the perf boost is that important, so just use `sortBy`

**Comment by 23Skidoo** [comment]
Closing in favour of #3969.

**Pull request closed without merging**